### PR TITLE
add Cronjob feature request #1795

### PR DESCRIPTION
**Pull request by realgam3**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add CronJob  Labels:
* `kompose.controller.type`: cronjob
  * `kompose.cronjob.schedule`: "5 4 * * *"
  * `kompose.cronjob.concurrency_policy`: (Allow|Forbid|Replace)

#### Which issue(s) this PR fixes:
Feature https://github.com/kubernetes/kompose/issues/1795

### CODE REVIEW

**Comment by realgam3** [comment]
@cdrage @AhmedGrati 
I think it's in a really good state,
I have added documentation and removed cronjob service type because it's more accurate to act as cronjob is like a pod.
I have added backoff limit so it will be possible to make cronjob run and die if it didn't succeed

We can check if schedule is correct by parsing it with library like: https://github.com/robfig/cron
Or by implementing a cronjob parser but I think it's unnecessary (Tell me if it's needed).

**Comment by realgam3** [comment]
> @cdrage @AhmedGrati I think it's in a really good state, I have added documentation and removed cronjob service type because it's more accurate to act as cronjob is like a pod. I have added backoff limit so it will be possible to make cronjob run and die if it didn't succeed
> 
> We can check if schedule is correct by parsing it with library like: https://github.com/robfig/cron Or by implementing a cronjob parser but I think it's unnecessary (Tell me if it's needed).
> 
> I had a conflict but I can rebase and force push so it will be one commit. (you can also squeeze if you like)

Any thoughts?
/joke

**Comment by cdrage** [comment]
Sorry been out on a work trip.

I will be reviewing this too by next week, as well as get a release out and fix our examples.

Many thanks for this awesome contribution! I've been wanting Cronjob added for a whileeeee.

**Comment by realgam3** [comment]
> Sorry been out on a work trip.
> 
> 
> 
> I will be reviewing this too by next week, as well as get a release out and fix our examples.
> 
> 
> 
> Many thanks for this awesome contribution! I've been wanting Cronjob added for a whileeeee. So it's amazing to see this in!

Wow thank you 😁
I wasn't sure you want to add this feature at all but I tried anyway 😂
I will continue to contribute, I'm using kompose in every CTF project 😁 so far I had 6+ CTFs with kompose in 3 years.

File: pkg/transformer/openshift/openshift.go
Comment (AhmedGrati) [inline pkg/transformer/openshift/openshift.go:334]: Same as above.

**Review by AhmedGrati — CHANGES_REQUESTED** [review]
I left some comments below, and I think that you need to add some unit tests.

File: pkg/transformer/kubernetes/kubernetes.go
Comment (realgam3) [reply pkg/transformer/kubernetes/kubernetes.go:498]: Copy paste error 😅
I'm embarrassed

File: pkg/loader/compose/compose.go
Comment (realgam3) [reply pkg/loader/compose/compose.go:776]: I've done it because of the other conversion errors that used the same message, but you are right

File: pkg/loader/compose/compose.go
Comment (realgam3) [reply pkg/loader/compose/compose.go:714]: I thought about it but I don't want to set a default if kubernetes set the new default to 10 we will need to change it, null will get removed because of the new feature that remove nulls. So this key will be removed if you didn't set it 😅

File: pkg/transformer/kubernetes/kubernetes.go
Comment (realgam3) [reply pkg/transformer/kubernetes/kubernetes.go:1623]: You're right but in the code we are checking types and conversion only at the end, and that's the reason there are 2 different checks in openshift and k8s, I think we should move everything to compose 😅

File: pkg/loader/compose/compose.go
Comment (realgam3) [reply pkg/loader/compose/compose.go:698]: It will allow you to write case insensitive but it return constant

File: pkg/loader/compose/compose.go
Comment (realgam3) [reply pkg/loader/compose/compose.go:706]: In returning empty string only because it will be removed by the new feature

File: pkg/loader/compose/compose.go
Comment (realgam3) [reply pkg/loader/compose/compose.go:768]: Absolutely right

**Comment by realgam3** [comment]
> I left some comments below, and I think that you need to add some unit tests.

I saw the comments? I will fix most of it asap and we will do the next round, I replied of some comments. I feel that if we change convention we will need to change it across the board for all other service types 😅

What do you think we need to test in the unit test? I will check other unit tests that is similar.
I've created end to end test.

File: pkg/loader/compose/compose.go
Comment (AhmedGrati) [reply pkg/loader/compose/compose.go:714]: Hmm, yeah I got your point. Actually, Kompose users are mainly Kubernetes beginners, and I believe that having the default attribute of `backoffLimit` already set to default can make users aware of it which is beneficial. This is my POV, leaving the final decision to @cdrage.

File: pkg/loader/compose/compose.go
Comment (AhmedGrati) [reply pkg/loader/compose/compose.go:706]: See the comment below.

**Comment by AhmedGrati** [comment]
> What do you think we need to test in the unit test? I will check other unit tests that is similar.

Unit tests of labels handler in `compose.go`, as well as tests for the Cronjob generation in `kubernetes.go` and `openshift.go`.

**Comment by realgam3** [comment]
I've done some changes from your comments @AhmedGrati,
Some code I keep the same because I want to keep the same code convention.
But I have made comments that show why I took that decisions.

feel free to tell me to change it anyway... @AhmedGrati @cdrage

File: pkg/loader/compose/compose.go
Comment (AhmedGrati) [reply pkg/loader/compose/compose.go:777]: got your point, but I see that we can take a step towards improving our error messages.

File: pkg/loader/compose/compose.go
Comment (cdrage) [reply pkg/loader/compose/compose.go:714]: TBH I agree with @AhmedGrati let's leave it as the Kubernetes default as that'd be best for conversion / learning for the user.

File: pkg/loader/compose/compose.go
Comment (cdrage) [reply pkg/loader/compose/compose.go:698]: agree on changing it, we shouldn't be allowing non-casesensitive, so I'm good with what @AhmedGrati said :)

**Review by cdrage — CHANGES_REQUESTED** [review]
This looks great! Only a few comments as @AhmedGrati has addressed the rest, but great job on adding cronjob!

**Comment by realgam3** [comment]
> This looks great! Only a few comments as @AhmedGrati has addressed the rest, but great job on adding cronjob!

Thank you 😁
There are some comments that need your involvement 😅

### We have some conventions like: * wrapping general error of the function that handles conversion. 
* failing from template generation with a general error that the creation of the manifest template failed.
* no spaces in some groups of label declarations.

And the most important one is setting null value to be cleared from the template instead of a default value taken from the kubernetes documentation there are good and bad things about both approaches and we need to decide 😅

**Comment by cdrage** [comment]
> > This looks great! Only a few comments as @AhmedGrati has addressed the rest, but great job on adding cronjob!
> 
> Thank you 😁 There are some comments that need your involvement 😅
> 
> ### We have some conventions like:
> * wrapping general error of the function that handles conversion.
> * failing from template generation with a general error that the creation of the manifest template failed.
> * no spaces in some groups of label declarations.
> 
> And the most important one is setting null value to be cleared from the template instead of a default value taken from the kubernetes documentation there are good and bad things about both approaches and we need to decide 😅

No problem.

To be honest those three points should be opened as separate issues as I believe them to be outside the scope of this PR.

- Our general error output sucks at the moment
- Failing with a general error isn't implemented yet
- No spaces should exist.

**Comment by realgam3** [comment]
> > > This looks great! Only a few comments as @AhmedGrati has addressed the rest, but great job on adding cronjob!
> 
> > 
> 
> > Thank you 😁 There are some comments that need your involvement 😅
> 
> > 
> 
> > ### We have some conventions like:
> 
> > * wrapping general error of the function that handles conversion.
> 
> > * failing from template generation with a general error that the creation of the manifest template failed.
> 
> > * no spaces in some groups of label declarations.
> 
> > 
> 
> > And the most important one is setting null value to be cleared from the template instead of a default value taken from the kubernetes documentation there are good and bad things about both approaches and we need to decide 😅
> 
> 
> 
> No problem.
> 
> 
> 
> To be honest those three points should be opened as separate issues as I believe them to be outside the scope of this PR.
> 
> 
> 
> - Our general error output sucks at the moment
> 
> - Failing with a general error isn't implemented yet
> 
> - No spaces should exist.

Got it so I will only fix the 2 spaces (lines).
I will fail if the input is insensitive like aLlOw but should if fail with "did you mean message or just fail?
What about the defaults should we set backoffRestart to 6 and concurrencyPolicy to Always or keep it null and kubernetes / openshift will set the defaults

**Comment by cdrage** [comment]
> > > > This looks great! Only a few comments as @AhmedGrati has addressed the rest, but great job on adding cronjob!
> > 
> > 
> > > 
> > 
> > 
> > > Thank you 😁 There are some comments that need your involvement 😅
> > 
> > 
> > > 
> > 
> > 
> > > ### We have some conventions like:
> > 
> > 
> > > * wrapping general error of the function that handles conversion.
> > 
> > 
> > > * failing from template generation with a general error that the creation of the manifest template failed.
> > 
> > 
> > > * no spaces in some groups of label declarations.
> > 
> > 
> > > 
> > 
> > 
> > > And the most important one is setting null value to be cleared from the template instead of a default value taken from the kubernetes documentation there are good and bad things about both approaches and we need to decide 😅
> > 
> > 
> > No problem.
> > To be honest those three points should be opened as separate issues as I believe them to be outside the scope of this PR.
> > 
> > * Our general error output sucks at the moment
> > * Failing with a general error isn't implemented yet
> > * No spaces should exist.
> 
> Got it so I will only fix the 2 spaces (lines). I will fail if the input is insensitive like aLlOw but should if fail with "did you mean message or just fail? What about the defaults should we set backoffRestart to 6 and concurrencyPolicy to Always or keep it null and kubernetes / openshift will set the defaults

Let Kubernetes / OpenShift set the defaults / don't include the keys when converting, and it'll automatically fill them in :)

Whatever is defaulted within the cluster, should be what's set

**Comment by realgam3** [comment]
@cdrage all fixed

File: docs/user-guide.md
Comment (cdrage) [inline docs/user-guide.md:211]: sorry to be a pain, can you give "example" inputs for the cronjobs, like:
```
      kompose.cronjob.schedule: "* * * * *"
      kompose.cronjob.concurrency_policy: "Forbid"
      kompose.cronjob.backoff_limit: "0"
```

Just put in brackets what they would possibly be (ex. "Forbig / Allow / Never"), etc.
    
Honestly we need to improve this table with a separate column with examples... But that'll be in a later PR

**Review by cdrage — CHANGES_REQUESTED** [review]
One small comment regarding documentation and this LGTM! After the documentation change I'm happy to merge this in and do a release (awesome new feature!)

File: pkg/loader/compose/compose.go
Comment (AhmedGrati) [reply pkg/loader/compose/compose.go:766]: @realgam3 any updates on this?

**Comment by realgam3** [comment]
@AhmedGrati added check if schedule is empty
@cdrage added examples in the documentation, I tried to make it as close as possible to other labels

**Comment by cdrage** [comment]
> @AhmedGrati added check if schedule is empty @cdrage added examples in the documentation, I tried to make it as close as possible to other labels

No problem! Just a quick rebase then against main and this LGTM!

**Comment by realgam3** [comment]
I'

> > @AhmedGrati added check if schedule is empty @cdrage added examples in the documentation, I tried to make it as close as possible to other labels
> 
> No problem! Just a quick rebase then against main and this LGTM!

I'm working on it, I have conflicts with the new documentation, but I hope I will fix it today.

**Pull request closed without merging**
